src/data/list/join.lean

```diff
@@ -148,4 +148,23 @@ begin
   { simp },
 end
 
+/-- We can rebracket `x ++ (l₁ ++ x) ++ (l₂ ++ x) ++ ... ++ (lₙ ++ x)` to
+`(x ++ l₁) ++ (x ++ l₂) ++ ... ++ (x ++ lₙ) ++ x` where `L = [l₁, l₂, ..., lₙ]`. -/
+lemma append_join_append (L : list (list α)) (x : list α) :
+  x ++ (list.map (λ l, l ++ x) L).join = (list.map (λ l, x ++ l) L).join ++ x :=
```

Comment:
Maybe it would be sensible to show that:
* `(list.map (λ l, l ++ x) L).join = x.intercalate l ++ x`
* `(list.map (λ l, x ++ l) L).join = x ++ x.intercalate l`
If you prove those two results, then this entire lemma follows from the above and `append_assoc`.

Reply:
Thanks for the suggestion, will do!

Reply:
Actually, it requires `L` to be non-empty. I am not sure it will be elegant that way.

Reply:
Good point, feel free to back out these lemmas from this PR.

Reply:
OK.